File: tests/test_optimization.cpp
#include "doctest.h"
#include "../include/environment.h"
#include <numeric>
using namespace std;

//*****************************
//*    Evaluation context     *
//****************************/

class ArithmeticContext : public EvaluationContext
{
public:
    int x{}, y{};

    ArithmeticContext(int px, int py) : x(px), y(py) {}

    int GetIntSemanticValue(int index) { return stoi(SemanticValue(index)); }
    int GetIntResult() { return stoi(result()); }
    void SetIntResult(int r) { result() = to_string(r); }
};

//*****************************
//*     Types declaration     *
//****************************/

enum Terms
{
    Var, Plus, Times, LeftParenthesis, RightParenthesis, // Terminals
    Expr, Term, Factor // NonTerminals
};

//*****************************
//*    Grammar declaration    *
//****************************/

// Term/Nonterm declaration.
const Terminal varTerm(Var, "var", { "x", "y", "1" });
const Terminal plusTerm(Plus, "Plus", { "+" });
const Terminal timesTerm(Times, "Times", { "*" });
const Terminal leftParenthesisTerm(LeftParenthesis, "LeftParenthesis", { "(" });
const Terminal rightParenthesisTerm(RightParenthesis, "RightParenthesis", { ")" });

const NonTerminal exprNonTerm(Expr, "EXPR");
const NonTerminal termNonTerm(Term, "TERM");
const NonTerminal factorNonTerm(Factor, "FACTOR");

// Grammar definition.
const ProductionRule rule1(
        exprNonTerm,
        { ProductionElement(exprNonTerm), ProductionElement(plusTerm), ProductionElement(termNonTerm) },
        [](EvaluationContext& ctx) {
            auto& arithmeticContext = dynamic_cast<ArithmeticContext&>(ctx);
            int n1 = arithmeticContext.GetIntSemanticValue(0);
            int n2 = arithmeticContext.GetIntSemanticValue(2);
            arithmeticContext.SetIntResult(n1 + n2);
        }
);

const ProductionRule rule2(
        exprNonTerm,
        { ProductionElement(termNonTerm) }
);

const ProductionRule rule3(
        termNonTerm,
        { ProductionElement(termNonTerm), ProductionElement(timesTerm), ProductionElement(factorNonTerm) },
        [](EvaluationContext& ctx) {
            auto& arithmeticContext = dynamic_cast<ArithmeticContext&>(ctx);
            int n1 = arithmeticContext.GetIntSemanticValue(0);
            int n2 = arithmeticContext.GetIntSemanticValue(2);
            arithmeticContext.SetIntResult(n1 * n2);
        }
);

const ProductionRule rule4(
        termNonTerm,
        { ProductionElement(factorNonTerm) }
);

const ProductionRule rule5(
        factorNonTerm,
        { ProductionElement(leftParenthesisTerm), ProductionElement(exprNonTerm), ProductionElement(rightParenthesisTerm) },
        1
);

const ProductionRule rule6(
        factorNonTerm,
        { ProductionElement(varTerm) },
        [](EvaluationContext& ctx) {
            auto& arithmeticContext = dynamic_cast<ArithmeticContext&>(ctx);
            string var = ctx.SemanticValue(0);

            int varValue;
            if (var == "x")
                varValue = arithmeticContext.x;
            else if (var == "y")
                varValue = arithmeticContext.y;
            else
                varValue = 1;

            arithmeticContext.SetIntResult(varValue);
        }
);

//*****************************
//*       Test routines       *
//****************************/
int target_func(int x, int y)
{
    return 1 + 2*x + y*y*y;
}

double fitness_function(SyntaxTree& solution)
{
    vector<int> diff;
    for (int x = 0; x <= 10; x++)
    {
        for (int y = 0; y <= 10; y++)
        {
            ArithmeticContext arithmeticContext(x, y);;
            solution.Evaluate(arithmeticContext);

            int solutionValue = arithmeticContext.GetIntResult();
            int expectedValue = target_func(x, y);

            diff.push_back(abs(solutionValue - expectedValue));
        }
    }

    int sum = std::accumulate(diff.begin(), diff.end(), 0);
    double error = (double) sum / (double) diff.size();

    return 1.0 / (1.0 + error);
}

TEST_CASE("Test individual evaluation")
{
    SyntaxTree removeParenthesisFrom(
            TreeNode(
                    rule5,
                    factorNonTerm,
                    {
                            TreeNode(leftParenthesisTerm, "("),
                            TreeNode(
                                    rule2,
                                    exprNonTerm,
                                    {
                                            TreeNode(
                                                    rule4,
                                                    termNonTerm,
                                                    {
                                                            TreeNode(
                                                                    rule6,
                                                                    factorNonTerm,
                                                                    {
                                                                            TreeNode(varTerm)
                                                                    })
                                                    })
                                    }),
                            TreeNode(rightParenthesisTerm, ")")
                    })
    );

    SyntaxTree removeParenthesisTo(
            TreeNode(
                    rule6,
                    factorNonTerm,
                    {
                            TreeNode(varTerm)
                    })
    );

    PruneRule removeParenthesis(removeParenthesisFrom, removeParenthesisTo);

    // GP Generator grammar
    SyntaxTree cst;
    EvaluationContext evaluationContext;
    Grammar grammar({ rule1, rule2, rule3, rule4, rule5, rule6 }, { removeParenthesis });

    grammar.CreateRandomTree(cst, 100);
    cst.PrintTree();

    cout << cst.SynthesizeExpression() << endl;
}

TEST_CASE("Test population optimization")
{
    SyntaxTree removeParenthesisFrom(
            TreeNode(
                    rule5,
                    factorNonTerm,
                    {
                            TreeNode(leftParenthesisTerm, "("),
                            TreeNode(
                                    rule2,
                                    exprNonTerm,
                                    {
                                            TreeNode(
                                                    rule4,
                                                    termNonTerm,
                                                    {
                                                            TreeNode(
                                                                    rule6,
                                                                    factorNonTerm,
                                                                    {
                                                                            TreeNode(varTerm)
                                                                    })
                                                    })
                                    }),
                            TreeNode(rightParenthesisTerm, ")")
                    })
    );

    SyntaxTree removeParenthesisTo(
            TreeNode(
                    rule6,
                    factorNonTerm,
                    {
                            TreeNode(varTerm)
                    })
    );

    PruneRule removeParenthesis(removeParenthesisFrom, removeParenthesisTo);
    Grammar grammar({ rule1, rule2, rule3, rule4, rule5, rule6 }, { removeParenthesis });
    Environment env(grammar, fitness_function, 200, 100, 5, 5, 0.4, RuntimeMode::MultiThread);

    cout << "Generation\t|\tScore\t|\tExpression" << endl;
    Population& lastGeneration = env.GetPopulation();
    Individual fittest = lastGeneration.GetFittestByRank(0);

    for (int i = 0; i < 30 && fittest.GetFitness() < 1; i++)
    {
        env.Optimize();
        fittest = lastGeneration.GetFittestByRank(0);
        cout << i << "\t|\t" << fittest.GetFitness() << "\t|\t" << fittest.GetExpression() << endl;
    }
}

File: python/test_optimization.py
import unittest
import gbgp
from enum import Enum


class ArithmeticContext(gbgp.EvaluationContext):
    x: int
    y: int

    def __init__(self, x: int, y: int):
        super().__init__()
        self.x = x
        self.y = y


class Terms(Enum):
    Var = 1
    Plus = 2
    Times = 3
    LeftParenthesis = 4
    RightParenthesis = 5
    Expr = 6
    Term = 7
    Factor = 8


varTerm = gbgp.Terminal(Terms.Var.value, "var", ["x", "y", "1"])
plusTerm = gbgp.Terminal(Terms.Plus.value, "Plus", ["+"])
timesTerm = gbgp.Terminal(Terms.Times.value, "Times", ["*"])
leftParenthesisTerm = gbgp.Terminal(Terms.LeftParenthesis.value, "LeftParenthesis", ["("])
rightParenthesisTerm = gbgp.Terminal(Terms.RightParenthesis.value, "RightParenthesis", [")"])

exprNonTerm = gbgp.NonTerminal(Terms.Expr.value, "EXPR")
termNonTerm = gbgp.NonTerminal(Terms.Term.value, "TERM")
factorNonTerm = gbgp.NonTerminal(Terms.Factor.value, "FACTOR")


def semantic_action1(ctx: gbgp.EvaluationContext):
    n1 = int(ctx.SemanticValue(0))
    n2 = int(ctx.SemanticValue(2))
    ctx.SetResult(str(n1 + n2))


rule1 = gbgp.ProductionRule(exprNonTerm,
                            [gbgp.ProductionElement(termNonTerm),
                             gbgp.ProductionElement(plusTerm),
                             gbgp.ProductionElement(termNonTerm)],
                            semantic_action1)

rule2 = gbgp.ProductionRule(exprNonTerm, [gbgp.ProductionElement(termNonTerm)])


def semantic_action3(ctx: gbgp.EvaluationContext):
    n1 = int(ctx.SemanticValue(0))
    n2 = int(ctx.SemanticValue(2))
    ctx.SetResult(str(n1 * n2))


rule3 = gbgp.ProductionRule(termNonTerm,
                            [gbgp.ProductionElement(termNonTerm),
                             gbgp.ProductionElement(timesTerm),
                             gbgp.ProductionElement(factorNonTerm)],
                            semantic_action3)
rule4 = gbgp.ProductionRule(termNonTerm, [gbgp.ProductionElement(factorNonTerm)])
rule5 = gbgp.ProductionRule(factorNonTerm,
                            [gbgp.ProductionElement(leftParenthesisTerm),
                             gbgp.ProductionElement(exprNonTerm),
                             gbgp.ProductionElement(rightParenthesisTerm)],
                            1)


def semantic_action6(ctx: gbgp.EvaluationContext):
    var = ctx.SemanticValue(0)
    if var == "x":
        var_value = ctx.x
    elif var == "y":
        var_value = ctx.y
    else:
        var_value = 1
    ctx.SetResult(str(var_value))


rule6 = gbgp.ProductionRule(factorNonTerm, [gbgp.ProductionElement(varTerm)], semantic_action6)
grammar = gbgp.Grammar([rule1, rule2, rule3, rule4, rule5, rule6])


def target_function(x: int, y: int) -> int:
    return 1 + 2 * x + y * y * y


def fitness_function(solution: gbgp.SyntaxTree) -> float:
    diff = []
    for x in range(9):
        for y in range(9):
            ctx = ArithmeticContext(x, y)
            solution.Evaluate(ctx)

            solution_value = int(ctx.GetResult())
            expected_value = target_function(x, y)
            diff.append(abs(solution_value - expected_value))

    error = sum(diff) / len(diff)
    return 1.0 / (1.0 + error)


class TestOptimization(unittest.TestCase):
    def test_population_optimization(self):
        env = gbgp.Environment(grammar, fitness_function, 200, 100, 5, 5, 0.4)
        last_generation = env.GetPopulation()
        fittest = last_generation.GetFittestByRank(0)

        i = 0
        while i < 30 and fittest.GetFitness() < 1:
            env.Optimize()
            last_generation = env.GetPopulation()
            fittest = last_generation.GetFittestByRank(0)
            print(str(i) + "\t|\t" + str(fittest.GetFitness()) + "\t|\t" + str(fittest.GetExpression()))
            i += 1

        self.assertEqual(fittest.GetFitness(), 1.0)


if __name__ == '__main__':
    unittest.main()
